File: autonav_ws/src/autonav_vision/src/expandify.py
#!/usr/bin/env python3

import rclpy
import numpy as np
import matplotlib.pyplot as plt
from enum import IntEnum

from nav_msgs.msg import MapMetaData, OccupancyGrid
from geometry_msgs.msg import Pose, Point
from cv_bridge import CvBridge
from autonav_libs import AutoNode, DeviceStateEnum as DeviceState

g_bridge = CvBridge()

g_mapData = MapMetaData()
g_mapData.width = 100
g_mapData.height = 100
g_mapData.resolution = 0.1
g_mapData.origin = Pose()
g_mapData.origin.position.x = -10.0
g_mapData.origin.position.y = -10.0


class Register(IntEnum):
    LOWER_HUE = 0
    LOWER_SATURATION = 1
    LOWER_VALUE = 2
    UPPER_HUE = 3
    UPPER_SATURATION = 4
    UPPER_VALUE = 5
    BLUR = 6
    BLUR_ITERATIONS = 7
    APPLY_FLATTENING = 8
    APPLY_REGION_OF_INTEREST = 9


class Expandifier(AutoNode):
    def __init__(self):
        super().__init__("autonav_vision_expandifier")

        self.verticalCameraRange = 2.75
        self.horizontalCameraRange = 3

        self.maxRange = 0.55
        self.noGoPercent = 0.75
        self.noGoRange = self.maxRange * self.noGoPercent
        
        self.maxRange = int(self.maxRange / (self.horizontalCameraRange / 80))
        self.noGoRange = int(self.noGoRange / (self.horizontalCameraRange / 80))

        self.metaData = MapMetaData(
            width=100,
            height=100,
            resolution=0.1,
            origin=Pose(
                position=Point(
                    x=-10.0,
                    y=-10.0
                )
            )
        )

    def setup(self):
        pts = list(range(-self.maxRange, self.maxRange + 1))
        self.circles = [(0, 0, 0)]
        for x in pts:
            for y in pts:
                if self.maxRange * self.noGoPercent <= np.sqrt(x ** 2 + y ** 2) < self.maxRange and (x + y) % 3 == 0:
                    self.circles.append((x, y, np.sqrt(x ** 2 + y ** 2)))

        self.m_rawConfigSubscriber = self.create_subscription(OccupancyGrid, "/autonav/cfg_space/raw", self.onRawCfgSpaceReceived, 1)
        self.m_configPublisher = self.create_publisher(OccupancyGrid, "/autonav/cfg_space/expanded", 1)
        self.setDeviceState(DeviceState.READY)
        self.setDeviceState(DeviceState.OPERATING)
        
    def onRawCfgSpaceReceived(self, grid: OccupancyGrid):
        cfg_space = [0] * (80 * 80)
        for x in range(80):
            for y in range(1, 80):
                if grid.data[x + y * 80] > 0:
                    for x_i, y_i, dist in self.circles:
                            index = (x + x_i) + 80 * (y + y_i)

                            if 0 <= (x + x_i) < 80 and 0 <= (y + y_i) < 80:
                                val_at_index = cfg_space[index]
                                linear_t = 100 - ((dist - self.noGoRange) /
                                                    (self.maxRange - self.noGoRange) * 100)

                                if dist <= self.noGoRange:
                                    # obstacle expansion
                                    cfg_space[index] = 100
                                elif dist <= 100 and val_at_index <= linear_t:
                                    # linearly decay
                                    cfg_space[index] = int(linear_t)

        new_msg = OccupancyGrid(data=cfg_space, info=self.metaData)
        self.m_configPublisher.publish(new_msg)
        
        # Show it using matplotlib
        # plt.imshow(np.array(cfg_space).reshape(80, 80))
        # plt.show(block = False)
        # plt.pause(0.01)

def main():
    rclpy.init()
    rclpy.spin(Expandifier())
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: autonav_ws/src/autonav_vision/src/expandify.cpp
#include "scr_core/node.h"
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose.hpp"
#include "nav_msgs/msg/map_meta_data.hpp"
#include "nav_msgs/msg/occupancy_grid.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "cv_bridge/cv_bridge.h"

namespace ExpandifyConstants
{
	const float VERTICAL_FOV = 2.75;
	const float HORIZONTAL_FOV = 3;
	const float MAP_RES_F = 80.0f;
	const int MAP_RES = 80;
}

struct Circle
{
	int x;
	int y;
	double radius;
};

class Expandify : public SCR::Node
{
public:
	Expandify() : SCR::Node("autonav_vision_expandifier") {}
	~Expandify() {}

	void configure() override
	{
		map = nav_msgs::msg::MapMetaData();
		map.width = 100;
		map.height = 100;
		map.resolution = 0.1;
		map.origin = geometry_msgs::msg::Pose();
		map.origin.position.x = -10.0;
		map.origin.position.y = -10.0;

		auto tempRange = maxRange * noGoPercent;
		maxRange = (int)(maxRange / (ExpandifyConstants::HORIZONTAL_FOV / ExpandifyConstants::MAP_RES_F));
		noGoRange = (int)(tempRange / (ExpandifyConstants::HORIZONTAL_FOV / ExpandifyConstants::MAP_RES_F));

		circles.push_back(Circle{0, 0, 0});
		for (int x = -maxRange; x <= maxRange; x++)
		{
			for (int y = -maxRange; y <= maxRange; y++)
			{
				if (maxRange * noGoPercent <= sqrt(x * x + y * y) && sqrt(x * x + y * y) < maxRange)
				{
					circles.push_back(Circle{x, y, sqrt(x * x + y * y)});
				}
			}
		}

		rawMapSubscriber = create_subscription<nav_msgs::msg::OccupancyGrid>("/autonav/cfg_space/raw", 20, std::bind(&Expandify::onConfigSpaceReceived, this, std::placeholders::_1));
		expandedMapPublisher = create_publisher<nav_msgs::msg::OccupancyGrid>("/autonav/cfg_space/expanded", 20);
		debugPublisher = create_publisher<sensor_msgs::msg::CompressedImage>("/autonav/cfg_space/expanded/image", 20);

		setDeviceState(SCR::DeviceState::OPERATING);
	}

	void transition(scr_msgs::msg::SystemState old, scr_msgs::msg::SystemState updated) override
	{
		if (updated.state == SCR::SystemState::AUTONOMOUS && getDeviceState() == SCR::DeviceState::READY)
		{
			setDeviceState(SCR::DeviceState::OPERATING);
		}

		if (updated.state != SCR::SystemState::AUTONOMOUS && getDeviceState() == SCR::DeviceState::OPERATING)
		{
			setDeviceState(SCR::DeviceState::READY);
		}
	}

	void onConfigSpaceReceived(const nav_msgs::msg::OccupancyGrid::SharedPtr cfg)
	{
		if (getDeviceState() != SCR::DeviceState::OPERATING || getSystemState().state != SCR::SystemState::AUTONOMOUS)
		{
			return;
		}

		performance.start("Expandification");

		std::vector<int8_t> cfg_space = std::vector<int8_t>(ExpandifyConstants::MAP_RES * ExpandifyConstants::MAP_RES);
		std::fill(cfg_space.begin(), cfg_space.end(), 0);

		for (int x = 0; x < ExpandifyConstants::MAP_RES; x++)
		{
			for (int y = 1; y < ExpandifyConstants::MAP_RES; y++)
			{
				if(cfg->data.at(x + y * ExpandifyConstants::MAP_RES) > 0)
				{
					for (Circle& circle : circles)
					{
						auto idx = (x + circle.x) + ExpandifyConstants::MAP_RES * (y + circle.y);
						auto expr_x = (x + circle.x) < ExpandifyConstants::MAP_RES && (x + circle.x) >= 0;
						auto expr_y = (y + circle.y) < ExpandifyConstants::MAP_RES && (y + circle.y) >= 0;
						if (expr_x && expr_y)
						{
							auto val = cfg_space.at(idx);
							auto linear = 100 - ((circle.radius - noGoRange) / (maxRange - noGoRange) * 100);

							if (circle.radius <= noGoRange)
							{
								cfg_space.at(idx) = 100;
							} else if (cfg_space.at(idx) <= 100 && val <= linear)
							{
								cfg_space.at(idx) = int(linear);
							}
						}
					}
				}
			}
		}

		auto newSpace = nav_msgs::msg::OccupancyGrid();
		newSpace.info = map;
		newSpace.data = cfg_space;
		expandedMapPublisher->publish(newSpace);

		cv::Mat image = cv::Mat(ExpandifyConstants::MAP_RES, ExpandifyConstants::MAP_RES, CV_8UC1, cfg_space.data());
		cv::cvtColor(image, image, cv::COLOR_GRAY2BGR);
		cv::resize(image, image, cv::Size(800, 800), 0, 0, cv::INTER_NEAREST);

		auto compressed = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", image).toCompressedImageMsg();
		compressed->header.stamp = this->now();
		compressed->header.frame_id = "map";
		compressed->format = "jpeg";
		debugPublisher->publish(*compressed);

		performance.end("Expandification");
	}

private:
	rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr rawMapSubscriber;
	rclcpp::Publisher<nav_msgs::msg::OccupancyGrid>::SharedPtr expandedMapPublisher;
	rclcpp::Publisher<sensor_msgs::msg::CompressedImage>::SharedPtr debugPublisher;

	nav_msgs::msg::MapMetaData map;
	
	float maxRange = 0.65;
	float noGoPercent = 0.70;
	int noGoRange = 0;
	std::vector<Circle> circles;
};

int main(int, char **)
{
	rclcpp::init(0, NULL);
	auto node = std::make_shared<Expandify>();
	rclcpp::spin(node);
	rclcpp::shutdown();
	return 0;
}